### ISSUE TITLE
Support delay signed .NET Core libraries

### ISSUE THREAD

**Issue by AndreyChechel** [obfuscar/obfuscar#288]
We faced an issue when Obfuscar doesn't sign .NET Core libraries when compilation is done in VS for Mac. The compilation process fails with the message `System.PlatformNotSupportedException: Strong-name signing is not supported on this platform.`. The reason is .NET Core lacking support of some [Reflection members](https://docs.microsoft.com/en-us/dotnet/core/compatibility/unsupported-apis#systemreflection), so there is no much can be done to get this round. 

But the next obvious step for us was to use delay signing, however, the similar error appeared. Below is the process we expected to work on Mac OS, but it fails on the step 2:
1. Build an assembly with the `DelaySign` option enabled.
2. Run Obfuscar GlobalTool to obfuscate the assembly and keep it "delay signed".
3. Run `sn.exe` outside of the build process to sign the obfuscated assembly with a private key.

The problem is the current way how Obfuscar detects a "delay signed" assembly - it is quite framework-implementation dependent. It was introduced in #164 (https://github.com/obfuscar/obfuscar/commit/610c4e37e2182248e7cba936acdbd8fcf2b9d8e5) by catching an `ArgumentException` [here](https://github.com/obfuscar/obfuscar/blob/d75dea5e0f0a47ca455eb7f6ff047c1cd98911c8/Obfuscar/Obfuscator.cs#L260-L266). And the failure point is [here](https://github.com/obfuscar/obfuscar/blob/d75dea5e0f0a47ca455eb7f6ff047c1cd98911c8/Obfuscar/Obfuscator.cs#L251).

The good thing is, if Obfuscar was told explicitly that it's a delayed signing, it'd be possible to avoid executing statements with non-supported `StrongNameKeyPair`, so the following block of code would be executed right away:

```cs
info.Definition.Name.PublicKey = keyPair;
info.Definition.Write(outName, parameters);
info.OutputFileName = outName;
```

Basically, another variable would be required: `DelaySign` of type `boolean`, that would instruct Obfuscar to keep an assembly "delay signed". We can submit a PR, if you're interested in these changes.

**Comment by lextm** [comment]
You should explore now whether you can sign the asssembly with Mono Cecil. The code base reports an exception because I didn't have time to investigate and at that time Mono Cecil didn't support that. Things might have changed already. (I really don't want to support delay signing as an option.)

**Comment by AndreyChechel** [comment]
> You should explore now whether you can sign the asssembly with Mono Cecil. 

Based on this  [comment](https://github.com/obfuscar/obfuscar/issues/164#issuecomment-440335703), it's possible.
There is also [a way to check](https://groups.google.com/forum/#!msg/mono-cecil/LyUwptVI_AA/OtSBXiCAOKwJ) whether a module is "delay signed" or not.

> I really don't want to support delay signing as an option.

If an explicit `DelaySign` variable is not a desired option, then I can propose throwing [this exception](https://github.com/obfuscar/obfuscar/blob/d75dea5e0f0a47ca455eb7f6ff047c1cd98911c8/Obfuscar/Obfuscator.cs#L270) for strong-named assemblies only (not for "delay signed" ones). In other words, if someone wants to obfuscate a "delay signed" assembly and leave it "delay signed" after obfuscation, they should not provide `KeyFile` or `KeyContainer` options - in this case Obfuscar will keep the existing Public Key without making any changes.

**Comment by AndreyChechel** [comment]
I see your idea now, I was mostly focused on the delay signing scenario. 
Confirming that the Mono Cecil implementation works on MacOS too.

As for the [PR](https://github.com/obfuscar/obfuscar/pull/291) I submitted, I still find it valid: if no public key is provided in the `obfuscar.xml` and the source assembly is delay signed, Obfuscar can still do its job without failing. Of course, with that Mono Cecil fix the scenario is no longer so critical.

**Comment by lextm** [comment]
So let's close this issue as fixed then.

I will review your pull request and see if it can be merged and shipped in the next release.
